### DISCUSSION TITLE
Acknowledgement after MessageListenerContainerRegistry.stop()

### DISCUSSION THREAD

**Discussion by MrLys** [awspring/spring-cloud-aws#885] [Q&A]
Hi!
Currently using spring-cloud-aws-sqs:3.0.2.
We're experiencing some issues when manually calling `MessageListenerContainerRegistry.stop()`. 
What we expect would happen (given `options.listenerShutdownTimeout(Duration.ofSeconds(30));`:
1. Messages are being received and processed my our `MessageListenerContainer`.
2. `MessageListenerContainerRegistry.stop()` is called.
3. The container will no longer poll for new messages.
4. The consumers will continue to run until done or until `Duration.ofSeconds(30)` has passed.
5. Those that succeed in time will `ack`, as they did finish successfully.
6. The `MessageListenerContainer` with finalize the shutdown.
7. System stops.

However, this is is not what is happening. The ack is not performed. 

So how can we stop processing new messages, but let the existing ones continue until `Duration.ofSeconds(30)` and ack once done?

**Comment by tomazfernandes** [comment]
@MrLys, you closed the discussion, am I right to assume it's working as expected?

Thanks.

**Reply by MrLys** [reply]
Hi @tomazfernandes!

Thank you for reaching out!
We realized that `acknowledgementShutdownTimeout` was set to be less than `listenerShutdownTimeout`, which likely was the reason we saw this.
However, while we're at it: After setting `acknowledgementShutdownTimeout` to 45s and `listenerShutdownTimeout` to 30s, we still see some events being processed twice. My investigations has shown that this should only happen if the container is shut down before the ack has been emitted, but after the system has finished processing the event. What is strange is that we're still seeing events being processed twice even though they are sub 500ms processing time. 
I can reopen this thread if you like.

**Reply by tomazfernandes** [reply]
Hi @MrLys, can you explain a bit further the scenario? What do you mean by "ack has been emitted"?

Perhaps setting log level for `BatchingAcknowledgementProcessor` to TRACE might help us understand whether a particular message is being batched for acknowledgement or not.

Thanks.
